lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -254,26 +279,140 @@ instance ((PersistPublicKey (key 'AccountK)), Typeable chain)
+        ccF = build $ toText $ role @chain
+        xpubF = hexF $ serializeXPub acct
+        acctF = prefixF 8 xpubF <> "..." <> suffixF 8 xpubF
+
+data VerificationKeyPool (key :: Depth -> * -> *) = VerificationKeyPool {
+      verPoolAccountPubKey
+        :: !(key 'AccountK XPub)
+    -- ^ Corresponding key for the pool (a pool is tied to only one account)
+    , verPoolGap
+        :: !AddressPoolGap
+    -- ^ The actual gap for the pool. This can't change for a given pool.
+    , verPoolIndexedKeys
+        :: !(Map KeyHash (Index 'Soft 'ScriptK, AddressState))
+    , verPoolKnownScripts :: !(Map ScriptHash [Index 'Soft 'ScriptK])
+        -- ^ Known script hashes that contain our verification key hashes
+        -- represented here by corresponding indices
+    } deriving (Generic)
+
+deriving instance (Show (key 'AccountK XPub), Show (key 'ScriptK XPub))
+    => Show (VerificationKeyPool key)
+
+deriving instance (Eq (key 'AccountK XPub), Eq (key 'ScriptK XPub))
+    => Eq (VerificationKeyPool key)
+
+instance (NFData (key 'AccountK XPub), NFData (key 'ScriptK XPub))
+    => NFData (VerificationKeyPool key)
+
+instance ((PersistPublicKey (key 'AccountK)))
+    => Buildable (VerificationKeyPool key) where
+    build (VerificationKeyPool acct (AddressPoolGap g) _ _) = mempty
+       <> ccF <> " " <> acctF <> " (gap=" <> build g <> ")\n"
+      where
+        ccF = build $ toText MultisigScript
         xpubF = hexF $ serializeXPub acct
         acctF = prefixF 8 xpubF <> "..." <> suffixF 8 xpubF
 
--- | Bring a 'AccountingStyle' type back to the term-level. This requires a type
+deriving instance Ord KeyHash
+
+deriveKeyHash
+    :: (WalletKey k, SoftDerivation k)
+    => k 'AccountK XPub
+    -> Index 'Soft 'ScriptK
+    -> KeyHash
+deriveKeyHash accXPub' =
+    toVerKeyHash . deriveAddressPublicKey accXPub' MultisigScript . coerce
+
+lookupKeyHash
+    :: KeyHash
+    -> VerificationKeyPool k
+    -> (Maybe (Index 'Soft 'ScriptK), VerificationKeyPool k)
+lookupKeyHash keyHash pool@(VerificationKeyPool accXPub g indexedHashKeys scripts) =
+    let updateKey (ix, _) = (ix, Used)
+    in case Map.lookup keyHash indexedHashKeys of
+        Nothing -> (Nothing, pool)
+        Just (ix, _) ->
+            ( Just ix
+            , VerificationKeyPool accXPub g
+                (Map.adjust updateKey keyHash indexedHashKeys) scripts)
+
+updateKnownScripts
+    :: (WalletKey k, SoftDerivation k)
+    => (Map ScriptHash [Index 'Soft 'ScriptK] -> Map ScriptHash [Index 'Soft 'ScriptK])
+    -> VerificationKeyPool k
+    -> VerificationKeyPool k
+updateKnownScripts trKnownScripts (VerificationKeyPool accXPub g indexedHashKeys knownScripts) =
+    let knownScripts' = trKnownScripts knownScripts
+    in unsafeVerificationKeyPool accXPub g indexedHashKeys knownScripts'
+
+toVerKeyHash
+    :: WalletKey k
+    => k (depth:: Depth) XPub
+    -> KeyHash
+toVerKeyHash = KeyHash . blake2b224 . xpubPublicKey . getRawKey
```

Comment:
Don't we only want `k 'ScriptK XPub` here?

Reply:
done in https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/bbdfbc8e50f6bc91fbb0be153a4a9467ea308ab5